keps/sig-api-machinery/5589-gogo-dependency/README.md

```diff
@@ -0,0 +1,638 @@
+will be forked to a location within the `kubernetes` or `kubernetes-sigs` project,
+and pruned / modified to only output the subset of generated code kept by `go-to-protobuf`.
+
+**Modify String() implementation**
+
+There is widespread reliance on Kubernetes API objects implementing `String()`,
+so preserving the existence of `String()` implementations is necessary to minimize disruption.
+
+However, this only *happens* to be done as part of protobuf generation,
+and the output is a pseudo-Go structure that requires large amounts of code to construct,
+and is not particularly meaningful from an API perspective.
+
+After the code-generation is forked and modifiable, the `String()` implementation will be
+switched to output a JSON encoding of the object. This will either be done by modifying
+the protobuf `String()` generator, or by creating a separate `String()` generator
+and disabling the protobuf `String()` generator.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+- `k8s.io/apimachinery/pkg/runtime/serializer/protobuf`: `2025-09-29` - `53.1%`
+
+##### Integration tests
+
+There is no feature enablement / disablement, configurable behavior, or runtime behavior change associated with this KEP.
+
+##### e2e tests
+
+There is no feature enablement / disablement, configurable behavior, or runtime behavior change associated with this KEP.
+
+### Graduation Criteria
+
+First phase:
+  * Complete steps 1-4: Removal of runtime use of gogo packages by API types and clients
+  * Result: No gogo packages are referenced from `k8s.io/api`, `k8s.io/apimachinery`, and `k8s.io/client-go` .go / go.mod files
+
+Second phase:
+  * Complete step 5: Removal of code-generation use of gogo packages
+  * Result: No gogo packages are referenced from any staging .go file, or as a direct dependency from any Kubernetes go.mod file
+
+Third phase:
+  * Complete step 6: modify `String()` implementations to be independent of protobuf generation
+  * Result: `String()` implementations for API objects can be generated independent of protobuf,
+    and return results 
+
+#### Deprecation
+
+n/a
+
+### Upgrade / Downgrade Strategy
+
+The changes proposed in this KEP do not modify built code or runtime behavior.
+There is no difference in serialized data from the client or server relative to other versions.
+
+### Version Skew Strategy
+
+The changes proposed in this KEP do not modify built code or runtime behavior.
+There is no difference in serialized data from the client or server relative to other versions.
```

Comment:
being pedantic, I think there is the version Skew  problem described in Risks and Mitigation, from code that start importing the new version kube APIs with that missing methods

Reply:
mm... using k8s.io/api at 1.x will also require k8s.io/apimachinery at 1.x, the versions will be matched, right?

Reply:
ok... I'm still not sure I understand the comment... is there something I could change to make this clearer or address the comment?